docs: :building_construction: design of `build()`, with diagram and pseudocode (#58)
seedcase-project/seedcase-flower: docs/design/interface/python.qmd

---
title: Python functions
---

This document describes the Python functions that form the CLI and
non-CLI functions. Non-CLI functions are those used by the CLI
functions. We include function signatures, documentation, some

> pseudocode, and diagrams that represent its interface along with

Reviewer: ```suggestion
pseudocode, and diagrams that represent the interface along with
```
?

explanations for these design decisions. It primarily serves as a
reference for us to develop from and track what has been finished and
what remains to be done.

We use symbols to indicate parts of the design that are actively being
worked on, done, or planned (see table below). For work that is planned
or is in progress, we include the function signatures and docstrings to
clarify the design. Once the interface is implemented (done), we will
remove the signatures from the documentation and point to the reference
documentation instead. The symbols we use are described in the table
below.

| Status | Description |
|:-----------------------------:|:----------------------------------------|
| {{< var done >}} | Interface that has been implemented. |
| {{< var wip >}} | Interface that is currently being worked on. |

> | {{< var planned >}} | Interface that is planned, but isn't being worked on currently. |

Reviewer 2: Didn't we have an includes for this at some point?

: A table showing the symbols used to indicate the status of interface
components, along with their descriptions.

## Overview

There are two parts to the Python interface: those functions that
represent the interface via the Terminal (CLI) and those functions that
are used by the CLI functions to implement their behaviour. We will call
them "CLI" and "non-CLI" functions. We have this split for a few
reasons:

-   Creates a clear separation between the code that is exposed through
    the CLI and code that could be used on its own as a Python package
    library.
-   Can apply functional programming practices to the non-CLI functions,
    as CLI functions often require the use of side effects and other
    non-functional practices.
-   More easily write unit tests for the non-CLI code, while write
    integration tests for the CLI code.
-   Keep the size of the CLI functions small, as they will be built up
    by functional units of non-CLI code.

## Python CLI functions

These functions are the direct Python equivalents to the CLI commands
described in the [CLI design page](cli.qmd). They are `build()` and
`view()`. This page describes the Python design and decisions made and

> won't repeat the content found in the CLI design page.

Reviewer: ```suggestion
These functions are the direct Python equivalents of the CLI commands
described on the [CLI design page](cli.qmd): `build()` and
`view()`. This page describes the Python design and decisions made, and
does not repeat the content of the CLI design page.
```

### {{< var planned >}} `build()`

> As already described in the CLI design page, this function has two

Reviewer: ```suggestion
As already described in the CLI design page, this function has four
```
?

parameters:

-   `uri` for the URI argument. The default is the `datapackage.json`
    file found in the working directory. See the [CLI design](cli.qmd)
    for reasons why we use a URI.
-   `style` for the style of output to use, which dictates the output
    file type and the number of files that are created. Either
    `"custom"` or one of the built-in styles.
-   `output_dir` to specify where the starting directory is for the
    output files. Defaults to the current working directory.
-   `quiet` to suppress output messages.

<!-- TODO: Or maybe instead of `quiet` do `verbose` so the default is quiet? -->

> The internal flow of this function is approximately represented in the

Reviewer: ```suggestion
The internal flow of this function is shown in the
```

flowchart TD
    uri:::input --> resolve_uri{{"resolve_uri()"}} --> path
    path --> load_config{{"load_config()"}} --> Config
    style_cfg[style]:::input --> load_config
    path --> read_dp{{"read_datapackage()"}} --> metadata
    metadata & Config --> build_sections{{"build_sections()"}} --> output["list[BuiltSection]"]
    output --> write_sections{{"write_sections()"}}
    output_dir:::input --> write_sections
    write_sections --> files["list[Path]"]
    files --> cli_message{{"cli_message()"}} --> stdout
    quiet:::input --> cli_message

    classDef input fill:#f96
```

::: {.callout-tip collapse="true" icon="false"}
## Pseudocode: `build()`

``` {.python filename="src/seedcase_flower/cli.py"}
from typing import Optional
from seedcase_flower.config import Style
from pathlib import Path
from dataclasses import dataclass

# To keep track of path and built content for each section.
@dataclass(frozen=True)
class BuiltSection:
    output_path: Path
    rendered_content: str

# TODO: Will need to convert them to Typer CLI args.
# See https://typer.tiangolo.com/tutorial/arguments/default/
# and https://typer.tiangolo.com/tutorial/commands/help/#rich-markdown-and-markup

> # CLI functions don't need return type.

Reviewer 2: ```suggestion
# CLI functions don't need a return type.
```

def build(
    uri: str = "datapackage.json",
    style: Optional[Style] = None,
    output_dir: Path = Path("."),
    quiet: bool = False
    ):
    """Build the human-friendly documentation from the `datapackage.json` file.

    Args:
        uri: The URI to the datapackage.json file. Defaults to
            `datapackage.json` in the current working directory.
        style: The style of output to use. Either one of the built-in styles
            `Style` or None. If None, it will look for a config file in
            the same directory as the `datapackage.json` file. If a config file is
            not found, it will use the default style (`Style.quarto_one_page`).
        output_dir: The directory to output the generated files to.
            Defaults to the current working directory.
        quiet: If True, suppress output messages.

    Returns:

>         Outputs a message of the files created (if quiet is not set).

Reviewer 3: Yea, I'm not sure what we would like the final output to be. Not just paths, but something maybe like what `quarto render` shows? Like a described list of the files created like:

```
:heavy_check_mark: Created `path/to/file.qmd`
```

Or output a tree structure of all created files? :thinking:

    """
    # Potential implementation steps:

    # Output maybe str? Path?
    # Use `match` inside for strictness on URI types?
    path: str = resolve_uri(uri)

    # TODO: Use match inside of load_config to have strict checks?
    # If style is None and no config path found, use default style.
    # If style is None, look for config file in same dir as datapackage.json.

>     # If style is provided, use that, not the config file.

Reviewer: ```suggestion
    # If style is provided, use that, not the config file.
    # If style is None, look for config file in same dir as datapackage.json.
    # If style is None and no config path found, use default style.
```

    config: Config = load_config(style, path)

    # Able to read from URI, e.g. `https` or `file` or `gh`
    metadata: dict = read_datapackage(path)

    # One item per section, rendered from template.
    # Internally uses Jinja2 to render templates with metadata
    output: list[BuiltSection] = build_sections(

>         metadata, # or properties as name?

Reviewer 2: ```suggestion
    properties: dict = read_datapackage(path)

    # One item per section, rendered from template.
    # Internally uses Jinja2 to render templates with metadata
    output: list[BuiltSection] = build_sections(
        properties, 
```
In the naming glossary on the architecture page, we define properties as: 
> Metadata fields (name-value pairs) of a descriptor loaded as a Python dictionary. This can be a subset of the original descriptor or the entire structure.

And we actually don't define what is meant by metadata - we've mostly used it for simplicity when describing Flower. Using properties here also aligns with how we use it in Sprout.

## {{< var wip >}} Python Interface

All the command line functionality can be called from Python. A few
examples are given below. Note that all examples assume the following
import convention:

```python
import seedcase_flower as fl
```

### Examples

```python

> fl.view("https://...")

Author: Note the lack of path= since path is an argument rather than an option on the command line. See note in that section for more details.

```
Views the metadata for a Data Package located online in the terminal

---

```python
fl.build()
```

Builds a documentation site in qmd format for the data package in the
current working directory.

---

```python
fl.build(
    "my-folder/datapackage.json",
    config=fl.Config(
        style="quarto-resource-listings",
        ci="quarto"

>     )

Author: Do we need to expose this config class to the user? Even if we need it on the backend (do we?) it seems like it only adds complexity to someone using flower. We already have a distinction between CLI arguments and options in Python: the former are positional parameters and the latter are named parameters, is that enough?

>     )

Reviewer 3: I've been writing some pseudocode up and I think the CLI/entry point Python code will be quite different from the non-CLI code, mainly because CLI code is mostly non-functional (only used for the side effect of creating files), while the non-CLI will need to be functional in nature. 

And yes, the Config class should have a one-to-one representation with the config file.
